src/process/ProcessManagerImpl.cpp

```diff
@@ -291,83 +306,21 @@ ProcessManagerImpl::handleSignalWait()
     {
         return;
     }
-    std::lock_guard<std::recursive_mutex> guard(gImplsMutex);
     for (;;)
     {
         int status = 0;
         int pid = waitpid(-1, &status, WNOHANG);
         if (pid > 0)
         {
-            auto pair = gImpls.find(pid);
-            if (pair == gImpls.end())
+            std::lock_guard<std::recursive_mutex> guard(gManagersMutex);
+            for (ProcessManagerImpl* manager : gManagers)
```

Comment:
Are you sure about this loop?
I would imagine that because each `ProcessManagerImpl` registered `mSigChild` they should have the callback triggered.
Or is asio kinda broken and follows a "last handler wins" type of contract?

Reply:
You can look at the "quick restart" test https://github.com/stellar/stellar-core/blob/master/src/herder/HerderTests.cpp#L903
that creates a simulation object (creates a bunch of apps with their own timer). This is the correct way to spin up multiple applications.

In some other places, 2 apps are created with the same clock.
There is a potential problem on shutdown as `ApplicationImpl::shutdownMainIOService()` kills the IOService (shared by multiple apps).
https://github.com/stellar/stellar-core/blob/master/src/overlay/OverlayTests.cpp#L67